env/expand.go

```diff
@@ -0,0 +1,195 @@
+	SetAction
+	// UnsetAction is an action to clear (if existing) an environment variable, like os.Unsetenv
+	UnsetAction
+	// SkipAction means that no action is performed (usually for an env with an empty value)
+	SkipAction
+)
+
+// Command describes an action performed on an envrionment variable
+type Command struct {
+	Action   Action
+	Variable Variable
+}
+
+// Variable is an environment variable
+type Variable struct {
+	Key   string
+	Value string
+	// IsSensitive is true if variable is marked (optionally) sensitive initally (for example a sensitive input or a secret),
+	// or recursively references any variable marked as sensitive.
+	// The goal is to keep track of any references secrets, so these can be redacted easily.
+	IsSensitive bool
+}
+
+// DeclarationSideEffects is returned by GetDeclarationsSideEffects()
+type DeclarationSideEffects struct {
+	// CommandHistory is an ordered list of commands: when performed in sequence,
+	// will result in a environment that contains the declared env vars
+	CommandHistory []Command
+	// ResultEnvironment is returned for reference,
+	// it will equal the environment after performing the commands
+	ResultEnvironment map[string]Variable
+}
+
+// EnvironmentSource implementations can return an initial environment
+type EnvironmentSource interface {
+	GetEnvironment() map[string]Variable
+}
+
+// DefaultEnvironmentSource is a default implementation of EnvironmentSource, returns the current environment
+type DefaultEnvironmentSource struct{}
+
+// GetEnvironment returns the current process' environment
+func (*DefaultEnvironmentSource) GetEnvironment() map[string]Variable {
+	processEnvs := os.Environ()
+	envs := make(map[string]Variable)
+
+	for _, env := range processEnvs {
+		key, value := SplitEnv(env)
+		if key == "" {
+			continue
+		}
+
+		envs[key] = Variable{
+			Key:         key,
+			Value:       value,
+			IsSensitive: false,
+		}
+	}
+
+	return envs
+}
+
+// SplitEnv splits an env returned by os.Environ
+func SplitEnv(env string) (key string, value string) {
+	const sep = "="
+	split := strings.SplitAfterN(env, sep, 2)
+	if split == nil {
+		return "", ""
+	}
+	key = strings.TrimSuffix(split[0], sep)
+	if len(split) > 1 {
+		value = split[1]
+	}
+	return
+}
+
+// GetDeclarationsSideEffects iterates over the list of ordered new declared variables sequentally and returns the needed
+// commands (like os.Setenv) to add the variables to the current environment.
+// The current process environment is not changed.
+// Variable expansion is done also, every new variable can reference the previous and initial environments (via EnvironmentSource)
+// The new variables (models.EnvironmentItemModel) can be defined in the envman definition file, or filled in directly.
+// If the source of the variables (models.EnvironmentItemModel) is the bitrise.yml workflow,
+// they will be in this order: App secrets; App level envs; Workflow level envs; Additional Step info envs; Input envs.
```

Comment:
what do you mean by `Additional Step info envs`

Reply:
Updated comment.